File: app/src/main/java/com/surycaty/tirartime/SelecionarJogadoresActivity.java
package com.surycaty.tirartime;

import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.RatingBar;
import android.widget.RelativeLayout;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;
import android.widget.Toast;

import com.surycaty.tirartime.dao.JogadorDAO;
import com.surycaty.tirartime.entidade.Jogador;
import com.surycaty.tirartime.entidade.JogadorTest;
import com.surycaty.tirartime.entidade.Time;
import com.surycaty.tirartime.util.Utils;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class SelecionarJogadoresActivity extends AppCompatActivity {

    private List<Jogador> jogadores = new ArrayList<Jogador>();
    private List<CheckBox> selecionados = new ArrayList<CheckBox>();

    private int qtdJogadores = 5;
    private EditText etNumJogadores;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_selecionar_jogadores);

        etNumJogadores = (EditText) findViewById(R.id.numJogadores);

        Button btnUpJogadores = (Button) findViewById(R.id.btnUpJogadores);
        btnUpJogadores.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(qtdJogadores < 11)
                    ++qtdJogadores;

                etNumJogadores.setText(String.valueOf(qtdJogadores));
            }
        });

        Button btnDownJogadores = (Button) findViewById(R.id.btnDownJogadores);
        btnDownJogadores.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(qtdJogadores > 2)
                    --qtdJogadores;

                etNumJogadores.setText(String.valueOf(qtdJogadores));
            }
        });


        //jogadores = JogadorTest.listaJogadores();
        JogadorDAO jogadorDao = new JogadorDAO(getApplicationContext());
        jogadores = jogadorDao.getListaJogadores();

        TableLayout table = (TableLayout) findViewById(R.id.idTableLayout);

        //Cabecalho da Tabela
        TableRow rowHeader = new TableRow(getApplicationContext());

        TextView tvNomeJogadorHeader = new TextView(this);
        tvNomeJogadorHeader.setTextColor(Color.BLACK);
        tvNomeJogadorHeader.setText("Atleta");
        rowHeader.addView(tvNomeJogadorHeader);

        TextView tvPosicao = new TextView(this);
        tvPosicao.setTextColor(Color.BLACK);
        tvPosicao.setText("Posição");
        rowHeader.addView(tvPosicao);

        TextView tvNivelHeader = new TextView(this);
        tvNivelHeader.setTextColor(Color.BLACK);
        tvNivelHeader.setText("Estatísticas");
        tvNivelHeader.setTextSize(18f);
        tvNivelHeader.setPadding(30, 30, 30, 30);

        rowHeader.addView(tvNivelHeader);
        rowHeader.setBackgroundColor(Color.rgb(200,200,200));

        table.addView(rowHeader);

        int top = 150;

        boolean isPar = false;
        boolean isPrimeira = true;

        for(Jogador j : jogadores) {

            CheckBox ck = new CheckBox(getApplicationContext());
            ck.setId(j.getId());
            ck.setText(j.getNome());
            ck.setTextSize(18f);
            ck.setTextColor(Color.BLACK);
            //ck.setBackgroundColor(Color.BLUE);
            ck.setPadding(30, 30, 30, 30);

            selecionados.add(ck);

            TableRow row = new TableRow(getApplicationContext());
            row.addView(ck);

            TextView textView = new TextView(this);
            textView.setTextColor(Color.BLACK);
            textView.setText("MC");
            textView.setPadding(20, 32, 20, 32);
            row.addView(textView);

            RatingBar ratingBar = new RatingBar(getApplicationContext());
            ratingBar.setNumStars(5);
            ratingBar.setRating((float)j.getLevel());
            ratingBar.setEnabled(false);
            ratingBar.setScaleY(0.5f);
            ratingBar.setScaleX(0.5f);;
            //ratingBar.setBackgroundColor(Color.RED);
            //ratingBar.setSecondaryProgress(R.drawable.ic_star_press);
            //ratingBar.setProgress(R.drawable.ic_star_press);
            //ratingBar.setBackgroundResource(R.drawable.ic_star);
            //ratingBar.setProgressDrawable(R.drawable.custom_ratingbar);


            row.addView(ratingBar);

            if(isPrimeira){
                isPar = true;
                isPrimeira = false;
            } else {
                if(isPar){
                    row.setBackgroundColor(Color.rgb(200,200,200));
                    isPar = false;
                } else {
                    row.setBackgroundColor(Color.WHITE);
                    isPar = true;
                }
            }

            row.setDividerPadding(10);

            table.addView(row);
        }

        Button btnTirarTime = (Button) findViewById(R.id.btnTirarTime);

        btnTirarTime.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {

                Toast mensagem = null;
                boolean hasErro = false;
                String textoErro = "";

                qtdJogadores = Integer.valueOf(etNumJogadores.getText().toString());

                if(qtdJogadores < 2 || qtdJogadores > 11) {

                    if(hasErro)
                        textoErro += "\n";

                    textoErro += "O Nº de jogadores tem que estar entre 2 e 11, favor corrigir!";

                    hasErro = true;
                }

                if(hasErro) {
                    mensagem = Utils.mensagem(getApplicationContext(), textoErro);
                    mensagem.show();;

                    return;
                }

                List<Jogador> lista = new ArrayList<Jogador>();

                for(CheckBox ck : selecionados) {
                    if(ck.isChecked()) {
                        for(Jogador j : jogadores) {
                            if(j.getId() == ck.getId())
                                lista.add(j);
                        }
                    }
                }

                if(lista.isEmpty() || (lista.size() < qtdJogadores * 2)) {
                    mensagem = Utils.mensagem(getApplicationContext(), "Número de jogadores insulficiente para formar times.\nFavor Verificar o Nº de jogadores por times");
                    mensagem.show();
                    return;
                }

                List<Time> times = sortearTimes(lista);
                Intent it = new Intent(SelecionarJogadoresActivity.this, TimesFormadosActivity.class);

                it.putExtra("timesFormados", (Serializable) times);

                startActivity(it);
            }
        });

    }

    private List<Time> sortearTimes(List<Jogador> lista) {

        List<Jogador> jogadores = lista;
        List<Time> times = new ArrayList<Time>();
        List<String> nomeTimes = JogadorTest.listaNomeTimes();

        int jogadoresPorTime = qtdJogadores;

        List<Integer> nivelOrdenado = new ArrayList<Integer>();

        for(int i = 5; i > 0; i--) {
            for(Jogador j : jogadores) {
                if(j.getLevel() == i) {
                    nivelOrdenado.add(j.getLevel());
                }
            }
        }

        int nivelAtual = 0;
        List<Integer> nivelFiltrado = new ArrayList<Integer>();

        for(Integer i : nivelOrdenado) {
            if(nivelAtual == 0) {
                nivelFiltrado.add(i);
            } else {
                if(i != nivelAtual) {
                    nivelFiltrado.add(i);
                }
            }
            nivelAtual = i;
        }

        Collections.shuffle(nivelFiltrado);

        int numTimes = jogadores.size() / jogadoresPorTime;

        for(int i = 0; i < jogadoresPorTime; i++){
            if(i == 0) {
                for (int t = 0; t < numTimes; ++t) {
                    Collections.shuffle(nomeTimes);
                    String nome = nomeTimes.get(0);
                    times.add(new Time(t, nome, new ArrayList<Jogador>()));
                    nomeTimes.remove(0);
                }

                if(jogadores.size() % jogadoresPorTime > 0) {
                    Collections.shuffle(nomeTimes);
                    String nome = nomeTimes.get(0);
                    times.add(new Time(times.size(), nome, new ArrayList<Jogador>()));
                }
            }

            for(Time time : times) {
                if(!nivelFiltrado.isEmpty())
                    nivelAtual = nivelFiltrado.get(0);

                int tam = jogadores.size();
                boolean hasNivel = false;

                for(int l = 0; l < tam; l++){
                    if(jogadores.get(l).getLevel() == nivelAtual) {
                        hasNivel = true;
                        break;
                    }
                }

                if(!hasNivel && nivelFiltrado.size() > 0) {
                    nivelFiltrado.remove(0);
                    if(nivelFiltrado.size() > 0) {
                        nivelAtual = nivelFiltrado.get(0);
                    }
                }

                Jogador jogadorSelecionado = selecionaJogador(jogadores, nivelAtual);

                if(jogadorSelecionado != null) {
                    time.getJogadores().add(jogadorSelecionado);

                    jogadores.remove(jogadorSelecionado);
                }
            }

        }

        if(numTimes < times.size()) {
            for(Time t : times) {
                if(t.getJogadores().size() < jogadoresPorTime) {
                    Jogador remove = times.get(times.size() - 1).getJogadores().get(0);
                    t.getJogadores().add(remove);

                    if(times.get(times.size() - 1).getJogadores().size() > 0) {
                        times.get(times.size() - 1).getJogadores().remove(remove);
                    }
                }
            }
        }

        /*System.out.println("Escalacao:");
        for(Time t : times) {
            System.out.println("Jogadores do Time: " + t.getNome());
            for(Jogador j : t.getJogadores()) {
                System.out.println("Nome: " + j.getNome() + " - Lvl: " + j.getLevel());
            }
            System.out.println("");
        }*/

        return times;
    }

    private static Jogador selecionaJogador(List<Jogador> jogadores,
                                            int nivelAtual) {

        List<Jogador> aux = new ArrayList<Jogador>();

        for(Jogador j : jogadores) {
            if(j.getLevel() == nivelAtual)
                aux.add(j);
        }

        if(aux.size() > 1) {
            Collections.shuffle(aux);
        }

        return (aux.isEmpty())? null : aux.get(0);
    }

}

File: app/src/main/java/com/surycaty/joga10/activity/SelecionarJogadoresActivity.kt
package com.surycaty.joga10.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.surycaty.joga10.R
import com.surycaty.joga10.dao.JogadorDAO
import com.surycaty.joga10.entidade.Jogador
import com.surycaty.joga10.entidade.JogadorTest
import com.surycaty.joga10.entidade.Time
import com.surycaty.joga10.util.Constantes
import com.surycaty.joga10.util.Utils
import kotlinx.android.synthetic.main.activity_selecionar_jogadores.*
import java.io.Serializable
import java.util.*
import kotlin.collections.ArrayList

class SelecionarJogadoresActivity : AppCompatActivity() {

    private var jogadores: List<Jogador> = ArrayList()
    private var adapter : SelecionarJogadoresActivity.CadastroJogadorAdapter? = null

    private var qtdJogadores = 5

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_selecionar_jogadores)

        val jogadorDao = JogadorDAO(this)
        jogadores = jogadorDao.listaJogadores

        this.adapter = SelecionarJogadoresActivity.CadastroJogadorAdapter(this@SelecionarJogadoresActivity, jogadores)

        lvSelecionarAtleta.adapter = this.adapter

        btnUpJogadores.setOnClickListener {
            if (qtdJogadores < 11)
                ++qtdJogadores

            numJogadores.setText(qtdJogadores.toString())

        }

        btnDownJogadores.setOnClickListener {
            if (qtdJogadores > 1)
                --qtdJogadores

            numJogadores.setText(qtdJogadores.toString())
        }

        btnTirarTime.setOnClickListener(View.OnClickListener {

            var hasErro = false
            var textoErro = ""

            qtdJogadores = Integer.valueOf(numJogadores.text.toString())

            if (qtdJogadores < 1 || qtdJogadores > 11) {

                if (hasErro)
                    textoErro += "\n"

                textoErro += Constantes.MSG_ERRO_NUMERO_JOGADORES_ENTRE_1_E_11

                hasErro = true
            }

            if (hasErro) {
                Utils.mensagem(applicationContext, textoErro).show()

                return@OnClickListener
            }

            val lista = ArrayList<Jogador>()

            for (sel in this.adapter!!.list) {
                if (sel.isSelecionado) {
                    lista.add(sel)
                }
            }

            if (lista.isEmpty() || lista.size < qtdJogadores * 2) {
                Utils.mensagem(applicationContext, Constantes.MSG_ERRO_NUMERO_JOGADORES_INSULFICIENTE).show()
                return@OnClickListener
            }

            val times = sortearTimes(lista)
            val it = Intent(this@SelecionarJogadoresActivity, TimesFormadosActivity::class.java)

            it.putExtra("timesFormados", times as Serializable)

            startActivity(it)
        })

    }

    private fun sortearTimes(lista: MutableList<Jogador>): List<Time> {

        val times = mutableListOf<Time>()
        val nomeTimes = JogadorTest.listaNomeTimes()

        val jogadoresPorTime = qtdJogadores

        val nivelOrdenado = mutableListOf<Int>()

        for (i in 5 downTo 1) {
            for (j in lista) {
                if (j.level == i) {
                    nivelOrdenado.add(j.level)
                }
            }
        }

        var nivelAtual = 0
        val nivelFiltrado = ArrayList<Int>()

        for (i in nivelOrdenado) {
            if (nivelAtual == 0) {
                nivelFiltrado.add(i)
            } else {
                if (i != nivelAtual) {
                    nivelFiltrado.add(i)
                }
            }
            nivelAtual = i
        }

        nivelFiltrado.shuffle()

        val numTimes = lista.size / jogadoresPorTime

        for (i in 0 until jogadoresPorTime) {
            if (i == 0) {
                for (t in 0 until numTimes) {
                    nomeTimes.shuffle()
                    val nome = nomeTimes[0]
                    times.add(Time(t, nome, ArrayList()))
                    nomeTimes.removeAt(0)
                }

                if (lista.size % jogadoresPorTime > 0) {
                    nomeTimes.shuffle()
                    val nome = nomeTimes[0]
                    times.add(Time(times.size, nome, ArrayList()))
                }
            }

            for (time in times) {
                if (!nivelFiltrado.isEmpty())
                    nivelAtual = nivelFiltrado[0]

                val tam = lista.size
                var hasNivel = false

                for (l in 0 until tam) {
                    if (lista[l].level == nivelAtual) {
                        hasNivel = true
                        break
                    }
                }

                if (!hasNivel && nivelFiltrado.size > 0) {
                    nivelFiltrado.removeAt(0)
                    if (nivelFiltrado.size > 0) {
                        nivelAtual = nivelFiltrado[0]
                    }
                }

                val jogadorSelecionado = selecionaJogador(lista, nivelAtual)

                if (jogadorSelecionado != null) {
                    time.jogadores!!.add(jogadorSelecionado)

                    lista.remove(jogadorSelecionado)
                }
            }

        }

        if (numTimes < times.size) {
            for (t in times) {
                if (t.jogadores!!.size < jogadoresPorTime) {
                    val remove = times[times.size - 1].jogadores!!.get(0)
                    t.jogadores!!.add(remove)

                    if (times[times.size - 1].jogadores!!.size > 0) {
                        times[times.size - 1].jogadores!!.remove(remove)
                    }
                }
            }
        }

        return times
    }

    private fun selecionaJogador(jogadores: List<Jogador>,
                                 nivelAtual: Int): Jogador? {

        val aux = ArrayList<Jogador>()

        for (j in jogadores) {
            if (j.level == nivelAtual)
                aux.add(j)
        }

        if (aux.size > 1) {
            aux.shuffle()
        }

        return if (aux.isEmpty()) null else aux[0]
    }

    class CadastroJogadorAdapter() : BaseAdapter() {

        var list : List<Jogador> = ArrayList()
        lateinit var context : Context

        constructor(context: Context, jogadores: List<Jogador>): this() {

            this.list = jogadores

            this.context = context

        }

        override fun getView(posicao: Int, view: View?, p2: ViewGroup?): View {
            var convertView : View? = view

            if(convertView == null) {
                convertView = View.inflate(context, R.layout.formar_times, null)
            }

            val jogador = this.getItem(posicao)

            val txtNomeJogador = convertView!!.findViewById(R.id.ckAtleta) as CheckBox
            val txtPosicaoJogador = convertView.findViewById(R.id.txtPosicaoSelecionar) as TextView
            val rtNivel = convertView.findViewById(R.id.rbNivelSelecionar) as RatingBar

            txtNomeJogador.text = jogador.nome
            txtNomeJogador.isChecked = jogador.isSelecionado
            txtPosicaoJogador.text = jogador.posicao
            rtNivel.rating = jogador.level.toFloat()

            txtNomeJogador.setOnClickListener {
                list.get(posicao).isSelecionado = txtNomeJogador.isChecked
            }

            return convertView
        }

        override fun getItem(p0: Int): Jogador {
            return list.get(p0)
        }

        override fun getItemId(p0: Int): Long {
            return p0.toLong()
        }

        override fun getCount(): Int {
            return list.size
        }

        internal var sp: Spinner? = null

        internal var posicoes = JogadorDAO.listaPosicoes

    }
}
